File: src/fluids/fluid_2d.cpp
#include "fluid_2d.h"
#include "../servers/rapier_physics_server_2d.h"
#include "../servers/rapier_project_settings.h"
#include <godot_cpp/classes/editor_settings.hpp>
#include <godot_cpp/classes/engine.hpp>
#include <godot_cpp/classes/project_settings.hpp>
#include <godot_cpp/classes/rendering_server.hpp>
#include <godot_cpp/classes/time.hpp>
#include <godot_cpp/classes/world2d.hpp>

RapierPhysicsServer2D *_get_rapier_physics_server() {
	auto *physics_server = dynamic_cast<RapierPhysicsServer2D *>(PhysicsServer2D::get_singleton());
	if (!physics_server) {
		ERR_PRINT_ONCE("Fluid2D node requires Rapier2D Physics Engine. Disabling it.");
	}
	return physics_server;
}

real_t Fluid2D::get_density() const {
	return density;
}

void Fluid2D::set_density(real_t p_density) {
	RapierPhysicsServer2D *rapier_physics_server = _get_rapier_physics_server();
	if (!rapier_physics_server) {
		return;
	}
	if (density != p_density) {
		density = p_density;
		rapier_physics_server->fluid_set_density(rid, density);
	}
}

void Fluid2D::set_lifetime(real_t p_lifetime) {
	lifetime = p_lifetime;
	// Delete older particles on process loop
	set_process(debug_draw || lifetime > 0.0);
}

real_t Fluid2D::get_lifetime() const {
	return lifetime;
}

PackedVector2Array Fluid2D::get_accelerations() const {
	RapierPhysicsServer2D *rapier_physics_server = _get_rapier_physics_server();
	if (!rapier_physics_server) {
		return PackedVector2Array();
	}
	return rapier_physics_server->fluid_get_accelerations(rid);
}
PackedVector2Array Fluid2D::get_velocities() const {
	RapierPhysicsServer2D *rapier_physics_server = _get_rapier_physics_server();
	if (!rapier_physics_server) {
		return PackedVector2Array();
	}
	return rapier_physics_server->fluid_get_velocities(rid);
}
PackedVector2Array Fluid2D::get_points() const {
	RapierPhysicsServer2D *rapier_physics_server = _get_rapier_physics_server();
	if (!rapier_physics_server) {
		return PackedVector2Array();
	}
	auto new_points = rapier_physics_server->fluid_get_points(rid);

	Transform2D gl_transform = get_global_transform();
	for (int i = 0; i < new_points.size(); i++) {
		new_points[i] = gl_transform.xform_inv(new_points[i]);
	}
	return new_points;
}
PackedFloat32Array Fluid2D::get_create_times() const {
	return create_times;
}
PackedVector2Array Fluid2D::create_rectangle_points(int width, int height) {
	PackedVector2Array new_points;
	new_points.resize(width * height);
	for (int i = 0; i < width; i++) {
		for (int j = 0; j < height; j++) {
			new_points[i + j * width] = Vector2(i * radius * 2, j * radius * 2);
		}
	}
	return new_points;
}
PackedVector2Array Fluid2D::create_circle_points(int p_radius) {
	PackedVector2Array new_points;
	for (float i = -p_radius; i <= p_radius; i += 1) {
		for (float j = -p_radius; j <= p_radius; j += 1) {
			float x = i * radius * 2;
			float y = j * radius * 2;
			if (Math::sqrt(i * i + j * j) <= p_radius) {
				new_points.append(Vector2(x, y));
			}
		}
	}
	return new_points;
}

void Fluid2D::set_points(PackedVector2Array p_points) {
	RapierPhysicsServer2D *rapier_physics_server = _get_rapier_physics_server();
	if (!rapier_physics_server) {
		return;
	}
	points = p_points;
	int old_times = create_times.size();
	create_times.resize(p_points.size());
	uint64_t ticks = Time::get_singleton()->get_ticks_msec();
	for (int i = old_times; i < p_points.size(); i++) {
		create_times[i] = ticks;
	}
	Transform2D gl_transform = get_global_transform();
	for (int i = 0; i < p_points.size(); i++) {
		p_points[i] = gl_transform.xform(p_points[i]);
	}
	rapier_physics_server->fluid_set_points(rid, p_points);
	queue_redraw();
}

void Fluid2D::add_points_and_velocities(PackedVector2Array p_points, PackedVector2Array p_velocities) {
	RapierPhysicsServer2D *rapier_physics_server = _get_rapier_physics_server();
	if (!rapier_physics_server) {
		return;
	}
	points.append_array(p_points);
	int old_times = create_times.size();
	uint64_t ticks = Time::get_singleton()->get_ticks_msec();
	for (int i = old_times; i < points.size(); i++) {
		create_times.append(ticks);
	}
	Transform2D gl_transform = get_global_transform();
	for (int i = 0; i < p_points.size(); i++) {
		p_points[i] = gl_transform.xform(p_points[i]);
	}
	rapier_physics_server->fluid_add_points_and_velocities(rid, p_points, p_velocities);
	queue_redraw();
}

void Fluid2D::set_points_and_velocities(PackedVector2Array p_points, PackedVector2Array p_velocities) {
	RapierPhysicsServer2D *rapier_physics_server = _get_rapier_physics_server();
	if (!rapier_physics_server) {
		return;
	}
	points = p_points;
	int old_times = create_times.size();
	create_times.resize(p_points.size());
	for (int i = old_times; i < p_points.size(); i++) {
		create_times[i] = Time::get_singleton()->get_ticks_msec();
	}
	Transform2D gl_transform = get_global_transform();
	for (int i = 0; i < p_points.size(); i++) {
		p_points[i] = gl_transform.xform(p_points[i]);
	}
	rapier_physics_server->fluid_set_points_and_velocities(rid, p_points, p_velocities);
	queue_redraw();
}

void Fluid2D::delete_points(PackedInt32Array p_indices) {
	RapierPhysicsServer2D *rapier_physics_server = _get_rapier_physics_server();
	if (!rapier_physics_server) {
		return;
	}
	rapier_physics_server->fluid_delete_points(rid, p_indices);
	for (int i = 0; i < p_indices.size(); i++) {
		points.remove_at(p_indices[i]);
		create_times.remove_at(p_indices[i]);
	}
	queue_redraw();
}

RID Fluid2D::get_rid() const {
	return rid;
}

void Fluid2D::set_effects(const TypedArray<FluidEffect2D> &p_effects) {
	RapierPhysicsServer2D *rapier_physics_server = _get_rapier_physics_server();
	if (!rapier_physics_server) {
		return;
	}
	effects = p_effects;
	rapier_physics_server->fluid_set_effects(rid, effects);
}

TypedArray<FluidEffect2D> Fluid2D::get_effects() const {
	return effects;
}

void Fluid2D::_bind_methods() {
	ClassDB::bind_method(D_METHOD("get_accelerations"), &Fluid2D::get_accelerations);
	ClassDB::bind_method(D_METHOD("get_velocities"), &Fluid2D::get_velocities);
	ClassDB::bind_method(D_METHOD("get_points"), &Fluid2D::get_points);
	ClassDB::bind_method(D_METHOD("get_create_times"), &Fluid2D::get_create_times);
	ClassDB::bind_method(D_METHOD("set_points", "points"), &Fluid2D::set_points);
	ClassDB::bind_method(D_METHOD("set_points_and_velocities", "points", "velocities"), &Fluid2D::set_points_and_velocities);
	ClassDB::bind_method(D_METHOD("add_points_and_velocities", "points", "velocities"), &Fluid2D::add_points_and_velocities);
	ADD_PROPERTY(PropertyInfo(Variant::PACKED_VECTOR2_ARRAY, "points"), "set_points", "get_points");

	ClassDB::bind_method(D_METHOD("get_density"), &Fluid2D::get_density);
	ClassDB::bind_method(D_METHOD("set_density", "density"), &Fluid2D::set_density);
	ADD_PROPERTY(PropertyInfo(Variant::FLOAT, "density", PROPERTY_HINT_RANGE, U"0,1,or_greater"), "set_density", "get_density");

	ClassDB::bind_method(D_METHOD("get_lifetime"), &Fluid2D::get_lifetime);
	ClassDB::bind_method(D_METHOD("set_lifetime", "lifetime"), &Fluid2D::set_lifetime);
	ADD_PROPERTY(PropertyInfo(Variant::FLOAT, "lifetime"), "set_lifetime", "get_lifetime");

	ClassDB::bind_method(D_METHOD("get_rid"), &Fluid2D::get_rid);

	ClassDB::bind_method(D_METHOD("get_effects"), &Fluid2D::get_effects);
	ClassDB::bind_method(D_METHOD("set_effects", "effects"), &Fluid2D::set_effects);
	ADD_PROPERTY(PropertyInfo(Variant::ARRAY, "effects", PROPERTY_HINT_TYPE_STRING,
						 String::num(Variant::OBJECT) + "/" + String::num(PROPERTY_HINT_RESOURCE_TYPE) + ":FluidEffect2D"),
			"set_effects", "get_effects");

	ClassDB::bind_method(D_METHOD("create_rectangle_points", "width", "height"), &Fluid2D::create_rectangle_points);
	ClassDB::bind_method(D_METHOD("create_circle_points", "radius"), &Fluid2D::create_circle_points);

	ClassDB::bind_method(D_METHOD("set_debug_draw", "debug_draw"), &Fluid2D::set_debug_draw);
	ClassDB::bind_method(D_METHOD("get_debug_draw"), &Fluid2D::get_debug_draw);
	ADD_PROPERTY(PropertyInfo(Variant::BOOL, "debug_draw"), "set_debug_draw", "get_debug_draw");
}

void Fluid2D::set_debug_draw(bool p_debug_draw) {
	if (debug_draw != p_debug_draw) {
		debug_draw = p_debug_draw;
		set_process(debug_draw || lifetime > 0.0);
		set_notify_transform(debug_draw);
		queue_redraw();
	}
}

bool Fluid2D::get_debug_draw() const {
	return debug_draw;
}

Fluid2D::Fluid2D() {
	RapierPhysicsServer2D *rapier_physics_server = _get_rapier_physics_server();
	if (!rapier_physics_server) {
		return;
	}
	rid = rapier_physics_server->fluid_create();
	radius = RapierProjectSettings::get_fluid_particle_radius();
}

Fluid2D::~Fluid2D() {
	RapierPhysicsServer2D *rapier_physics_server = _get_rapier_physics_server();
	if (!rapier_physics_server) {
		return;
	}
	rapier_physics_server->free_rid(rid);
}

void Fluid2D::_delete_old_particles() {
	if (lifetime <= 0.0) {
		return;
	}
	uint64_t ticks = Time::get_singleton()->get_ticks_msec();
	PackedInt32Array to_remove;
	for (int i = 0; i < create_times.size(); i++) {
		if (ticks - create_times[i] > lifetime * 1000) {
			to_remove.append(i);
		}
	}
	if (!to_remove.is_empty()) {
		to_remove.reverse();
		delete_points(to_remove);
	}
}

void Fluid2D::_notification(int p_what) {
	RapierPhysicsServer2D *rapier_physics_server = _get_rapier_physics_server();
	if (!rapier_physics_server) {
		return;
	}
	static int ticks = 0;
	switch (p_what) {
		case NOTIFICATION_PROCESS: {
			if (debug_draw) {
				queue_redraw();
			}
			ticks++;
			if (ticks % 60 == 0) {
				_delete_old_particles();
			}
		} break;
		case NOTIFICATION_ENTER_TREE:
		case NOTIFICATION_WORLD_2D_CHANGED:
		case NOTIFICATION_TRANSFORM_CHANGED:
		case NOTIFICATION_LOCAL_TRANSFORM_CHANGED:
		case NOTIFICATION_TRANSLATION_CHANGED: {
			RID space = get_world_2d()->get_space();
			rapier_physics_server->fluid_set_space(rid, space);
			set_points(points);
			if (debug_draw) {
				queue_redraw();
			}
		} break;

		case NOTIFICATION_EXIT_TREE: {
			rapier_physics_server->fluid_set_space(rid, RID());
		} break;
		case NOTIFICATION_DRAW: {
			if (debug_draw) {
				points = get_points();
				for (int i = 0; i < points.size(); i++) {
					draw_rect(Rect2(points[i] - Vector2(radius / 2.0, radius / 2.0), Vector2(radius, radius)), Color::named("WHITE"));
				}
			}
		} break;
	}
}

File: src/fluids/fluid_2d.rs
use godot::classes::Engine;
use godot::classes::notify::CanvasItemNotification;
use godot::prelude::*;

use super::fluid_impl::FluidImpl;
use crate::servers::RapierPhysicsServer;
use crate::servers::rapier_project_settings::RapierProjectSettings;
use crate::types::*;
#[derive(GodotClass)]
#[class(base=Node2D,tool)]
/// The fluid node. Use this node to simulate fluids in 2D.
pub struct Fluid2D {
    #[var(get)]
    pub(crate) rid: Rid,
    #[var(get)]
    pub(crate) radius: real,
    #[export]
    #[var(get, set = set_debug_draw)]
    pub(crate) debug_draw: bool,
    #[export]
    #[var(get, set = set_density)]
    pub(crate) density: real,
    #[export]
    #[var(get, set = set_lifetime)]
    pub(crate) lifetime: real,
    #[export]
    #[var(get, set = set_effects)]
    pub(crate) effects: Array<Option<Gd<Resource>>>,

    #[export]
    #[var(get = get_points, set = set_points)]
    pub(crate) points: PackedVectorArray,
    pub(crate) create_times: PackedFloat32Array,

    // add a group for collision layers/masks
    #[export_group(name = "Collision", prefix = "collision_")]
    #[export(flags_2d_physics)]
    #[var(get = get_collision_layer, set = set_collision_layer)]
    pub(crate) collision_layer: u32,
    #[export(flags_2d_physics)]
    #[var(get = get_collision_mask, set = set_collision_mask)]
    pub(crate) collision_mask: u32,
    base: Base<Node2D>,
}
#[godot_api]
impl Fluid2D {
    #[func]
    /// Set the points of the fluid particles.
    fn set_points(&mut self, points: PackedVectorArray) {
        FluidImpl::set_points(self, points);
        self.to_gd().queue_redraw();
    }

    #[func]
    /// Set the density of the fluid particles.
    fn set_density(&mut self, density: real) {
        FluidImpl::set_density(self, density);
    }

    #[func]
    /// Set the lifetime of the fluid particles.
    fn set_lifetime(&mut self, lifetime: real) {
        FluidImpl::set_lifetime(self, lifetime);
    }

    #[func]
    /// Set the debug draw of the fluid particles.
    fn set_debug_draw(&mut self, debug_draw: bool) {
        FluidImpl::set_debug_draw(self, debug_draw);
        self.to_gd().queue_redraw();
    }

    #[func]
    /// Get the accelerations of the fluid particles.
    fn get_accelerations(&self) -> PackedVectorArray {
        FluidImpl::get_accelerations(self)
    }

    #[func]
    /// Get the remaining times of the fluid particles. Use this to draw particles with less alpha as time runs out.
    fn get_remaining_times(&self) -> PackedFloat32Array {
        FluidImpl::get_remaining_times(self)
    }

    #[func]
    /// Get the velocities of the fluid particles.
    fn get_velocities(&self) -> PackedVectorArray {
        FluidImpl::get_velocities(self)
    }

    #[func]
    /// Get the points of the fluid particles.
    fn get_points(&self) -> PackedVectorArray {
        FluidImpl::get_points(self)
    }

    #[func]
    /// Create the points of the fluid particles inside a rectangle.
    fn create_rectangle_points(&self, width: i32, height: i32) -> PackedVectorArray {
        let mut new_points = PackedVectorArray::default();
        new_points.resize((width * height) as usize);
        for i in 0..width {
            for j in 0..height {
                new_points[(i + j * width) as usize] =
                    Vector2::new(i as f32 * self.radius * 2.0, j as f32 * self.radius * 2.0);
            }
        }
        new_points
    }

    #[func]
    /// Create the points of the fluid particles inside a circle.
    fn create_circle_points(&self, radius: i32) -> PackedVectorArray {
        let mut new_points = PackedVectorArray::default();
        for i in -radius..radius {
            for j in -radius..radius {
                let x = i as f32 * self.radius * 2.0;
                let y = j as f32 * self.radius * 2.0;
                if i * i + j * j <= radius * radius {
                    new_points.push(Vector2::new(x, y));
                }
            }
        }
        new_points
    }

    #[func]
    /// Add the points (and velocities) to the fluid particles.
    fn add_points_and_velocities(
        &mut self,
        points: PackedVectorArray,
        velocities: PackedVectorArray,
    ) {
        FluidImpl::add_points_and_velocities(self, points, velocities);
        self.to_gd().queue_redraw();
    }

    #[func]
    /// Set the points (and velocities) of the fluid particles.
    fn set_points_and_velocities(
        &mut self,
        points: PackedVectorArray,
        velocities: PackedVectorArray,
    ) {
        FluidImpl::set_points_and_velocities(self, points, velocities);
        self.to_gd().queue_redraw();
    }

    #[func]
    /// Delete the points of the fluid particles.
    fn delete_points(&mut self, indices: PackedInt32Array) {
        FluidImpl::delete_points(self, indices);
        self.to_gd().queue_redraw();
    }

    #[func]
    /// Set the effects of the fluid particles.
    fn set_effects(&mut self, effects: Array<Option<Gd<Resource>>>) {
        FluidImpl::set_effects(self, effects);
    }

    #[func]
    /// Get the collision mask of the fluid particles.
    fn get_collision_mask(&self) -> u32 {
        FluidImpl::get_collision_mask(self)
    }

    #[func]
    /// Set the collision mask of the fluid particles.
    fn set_collision_mask(&mut self, mask: u32) {
        FluidImpl::set_collision_mask(self, mask);
    }

    #[func]
    /// Get the collision layer of the fluid particles.
    fn get_collision_layer(&self) -> u32 {
        FluidImpl::get_collision_layer(self)
    }

    #[func]
    /// Set the collision layer of the fluid particles.
    fn set_collision_layer(&mut self, layer: u32) {
        FluidImpl::set_collision_layer(self, layer);
    }

    #[func]
    /// Get the indices of the fluid particles inside an AABB.
    fn get_particles_in_aabb(&self, aabb: Rect2) -> PackedInt32Array {
        FluidImpl::get_particles_in_aabb(self, aabb)
    }

    #[func]
    /// Get the indices of the fluid particles inside a circle.
    fn get_particles_in_circle(&self, center: Vector2, radius: real) -> PackedInt32Array {
        FluidImpl::get_particles_in_ball(self, center, radius)
    }
}
#[godot_api]
impl INode2D for Fluid2D {
    fn init(base: Base<Node2D>) -> Self {
        Self {
            rid: RapierPhysicsServer::fluid_create(),
            radius: RapierProjectSettings::get_fluid_particle_radius(),
            debug_draw: false,
            density: 1.0,
            lifetime: 0.0,
            effects: Array::new(),
            points: PackedVectorArray::new(),
            create_times: PackedFloat32Array::new(),
            collision_mask: 1,
            collision_layer: 1,
            base,
        }
    }

    fn on_notification(&mut self, p_what: CanvasItemNotification) {
        match p_what {
            CanvasItemNotification::PROCESS => {
                if self.debug_draw {
                    self.to_gd().queue_redraw();
                }
                if !Engine::singleton().is_editor_hint() {
                    FluidImpl::delete_old_particles(self);
                }
            }
            CanvasItemNotification::ENTER_TREE | CanvasItemNotification::WORLD_2D_CHANGED => {
                let mut space_rid = Rid::Invalid;
                if let Some(space) = self.to_gd().get_world_2d() {
                    space_rid = space.get_space();
                }
                let rid = self.rid;
                let guard = self.base_mut();
                RapierPhysicsServer::fluid_set_space(rid, space_rid);
                drop(guard);
                self.set_points(self.points.clone());
                let mut fluid_gd = self.to_gd();
                fluid_gd.set_notify_transform(self.debug_draw);
                fluid_gd.queue_redraw();
            }
            CanvasItemNotification::TRANSFORM_CHANGED
            | CanvasItemNotification::LOCAL_TRANSFORM_CHANGED
            | CanvasItemNotification::TRANSLATION_CHANGED => {
                // Only update debug draw, do not re-add points to Rapier (prevents duplication)
                let mut fluid_gd = self.to_gd();
                fluid_gd.set_notify_transform(self.debug_draw);
                fluid_gd.queue_redraw();
            }
            CanvasItemNotification::EXIT_TREE => {
                let rid = self.rid;
                let guard = self.base_mut();
                RapierPhysicsServer::fluid_set_space(rid, Rid::Invalid);
                drop(guard);
            }
            CanvasItemNotification::DRAW => {
                if self.debug_draw {
                    for point in self.get_points().as_slice() {
                        let mut color = Color::WHITE;
                        color.a = 0.4;
                        self.to_gd().draw_rect(
                            Rect2::new(
                                *point - Vector2::new(self.radius / 2.0, self.radius / 2.0),
                                Vector2::new(self.radius, self.radius),
                            ),
                            color,
                        );
                    }
                }
            }
            _ => {}
        }
    }
}
impl Drop for Fluid2D {
    fn drop(&mut self) {
        if self.rid != Rid::Invalid {
            PhysicsServer::singleton().free_rid(self.rid);
        }
    }
}
